### PR TITLE
fix(fulltext-index): single segment is not sufficient for >50M rows SST

### PR DESCRIPTION
**Pull request by zhongzc**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

When building the fulltext index, an oversized SST was encountered, which exceeded a single segment buffer, causing the index building to fail.

## What's changed and what's your intention?

Use `IndexWriter` to build the index to allow multiple segments for one SST.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

File: src/index/src/fulltext_index/create/tantivy.rs [hunk 1]
```diff
@@ -59,10 +68,17 @@ impl TantivyFulltextIndexCreator {
 
         let memory_limit = Self::sanitize_memory_limit(memory_limit);
 
-        let writer = SingleSegmentIndexWriter::new(index, memory_limit).context(TantivySnafu)?;
+        // Use one thread to keep order of the row id.
+        let writer = index
+            .writer_with_num_threads(1, memory_limit)
```
Comment (evenyag) [inline src/index/src/fulltext_index/create/tantivy.rs:73]: How many threads tantivy uses if we don't set it? Will it affect performance?

File: src/index/src/fulltext_index/create/tantivy.rs [hunk 2]
```diff
@@ -96,20 +112,25 @@ impl TantivyFulltextIndexCreator {
 impl FulltextIndexCreator for TantivyFulltextIndexCreator {
     async fn push_text(&mut self, text: &str) -> Result<()> {
         let writer = self.writer.as_mut().context(FinishedSnafu)?;
-        let doc = doc!(self.text_field => text);
-        writer.add_document(doc).context(TantivySnafu)
+        let doc = doc!(self.text_field => text, self.rowid_field => self.max_rowid);
```
Comment (evenyag) [inline src/index/src/fulltext_index/create/tantivy.rs:115]: Looks like we need more space to store the index now.

File: src/index/src/fulltext_index/create/tantivy.rs [hunk 3]
```diff
@@ -16,28 +16,36 @@ use std::path::Path;
...
-use crate::fulltext_index::error::{FinishedSnafu, IoSnafu, Result, TantivySnafu};
+use crate::fulltext_index::error::{FinishedSnafu, IoSnafu, JoinSnafu, Result, TantivySnafu};
 use crate::fulltext_index::{Analyzer, Config};
 
 pub const TEXT_FIELD_NAME: &str = "greptime_fulltext_text";
+pub const ROWID_FIELD_NAME: &str = "greptime_fulltext_rowid";
 
 /// `TantivyFulltextIndexCreator` is a fulltext index creator using tantivy.
 ///
 /// Here use a single segment to store the index so the maximum capacity for
```
Comment (evenyag) [inline src/index/src/fulltext_index/create/tantivy.rs:35]: This comment is outdated.

File: src/index/src/fulltext_index/create/tantivy.rs [hunk 4]
```diff
@@ -59,10 +68,17 @@ impl TantivyFulltextIndexCreator {
 
         let memory_limit = Self::sanitize_memory_limit(memory_limit);
 
-        let writer = SingleSegmentIndexWriter::new(index, memory_limit).context(TantivySnafu)?;
```
Comment (evenyag) [inline src/index/src/fulltext_index/create/tantivy.rs:62]: The comment above said that `the maximum capacity for the index is limited to 2<<31 rows (around 2 billion rows)`. Why do we get an error when there are 50M rows?

File: src/index/src/fulltext_index/search/tantivy.rs [hunk 5]
```diff
@@ -77,9 +79,48 @@ impl FulltextIndexSearcher for TantivyFulltextIndexSearcher {
...
+            .searchable_segment_metas()
+            .context(TantivySnafu)?;
+
+        // FAST PATH: only one segment, the doc id is the same as the row id.
+        //            Also for compatibility with the old version.
+        if seg_metas.len() == 1 {
+            return Ok(doc_addrs.into_iter().map(|d| d.doc_id).collect());
+        }
+
+        // SLOW PATH: multiple segments, need to calculate the row id.
```
Comment (evenyag) [inline src/index/src/fulltext_index/search/tantivy.rs:97]: If we limit the total rows in an SST file, can we get rid of this issue?

File: src/index/src/fulltext_index/create/tantivy.rs (hunk 1 above)
Comment (zhongzc) [reply src/index/src/fulltext_index/create/tantivy.rs:73]: When using `SingleSegmentIndexWriter`, one thread is used as well, and there is not much change in our tests.

File: src/index/src/fulltext_index/create/tantivy.rs (hunk 2 above)
Comment (zhongzc) [reply src/index/src/fulltext_index/create/tantivy.rs:115]: Judging from testing, there won’t be much difference

File: src/index/src/fulltext_index/create/tantivy.rs (hunk 4 above)
Comment (zhongzc) [reply src/index/src/fulltext_index/create/tantivy.rs:62]: An internal buffer has a size limit. `SingleSegmentIndexWriter` is a dragon API and does not pay attention to this size limit, resulting in an error after exceeding it.

File: src/index/src/fulltext_index/search/tantivy.rs (hunk 5 above)
Comment (zhongzc) [reply src/index/src/fulltext_index/search/tantivy.rs:97]: It can alleviate the problem, but there is no guarantee

**Review by evenyag — APPROVED** [review]

File: src/index/src/fulltext_index/create/tantivy.rs (hunk 4 above)
Comment (waynexia) [reply src/index/src/fulltext_index/create/tantivy.rs:62]: 🐉

Comment (evenyag) [reply src/index/src/fulltext_index/create/tantivy.rs:62]: 🐉

File: src/index/src/fulltext_index/search/tantivy.rs [hunk 6]
```diff
@@ -77,9 +79,48 @@ impl FulltextIndexSearcher for TantivyFulltextIndexSearcher {
...
+                // the offset is the same for all rows in the same segment.
+                let doc: TantivyDocument = searcher.doc(doc_addr).context(TantivySnafu)?;
+                let rowid = doc
+                    .get_first(rowid_field)
+                    .and_then(|v| v.as_u64())
+                    .context(TantivyDocNotFoundSnafu { doc_addr })?;
+
+                let offset = rowid as u32 - doc_addr.doc_id;
+                seg_offsets.insert(doc_addr.segment_ord, offset);
+                offset
```
Comment (evenyag) [inline src/index/src/fulltext_index/search/tantivy.rs:118]: I can't understand this code today. I used to understand it yesterday......

What's the relationship between `doc`, `doc_id`, `segment_ord`, and `row_id`?

Comment (waynexia) [reply src/index/src/fulltext_index/search/tantivy.rs:118]: Is the purpose of this paragraph (the offset cache) to reduce call to `doc.get_first(rowid_field)`?

Comment (zhongzc) [reply src/index/src/fulltext_index/search/tantivy.rs:118]: > What's the relationship between `doc`, `doc_id`, `segment_ord`, and `row_id`?

image

**Review by waynexia — APPROVED** [review]

File: src/index/src/fulltext_index/search/tantivy.rs (hunk 6 above)
Comment (waynexia) [reply src/index/src/fulltext_index/search/tantivy.rs:118]: Awesome 🤩

**Pull request merged**